### PR TITLE
Feat/280/deploy conf file for aws deployment config

### PR DESCRIPTION
**Pull request by ttmc**
`awsdeploy.sh` and `launch_ec2_nodes.py` now read an AWS deployment configuration file named `deploy_conf.py`, rather than having a ton of command-line options.

`deploy_conf.py` has a format that can be interpreted as a Bash script and as a Python script. The provided `deploy_conf.py` has many comments to explain each setting and to give example values.

`awsdeploy.sh` starts out by calling `validate_deploy_conf.py` to check the values in `deploy_conf.py`

I updated the BigchainDB documentation to explain how to use `awsdeploy.sh` now. I think it's actually gotten easier.

I also fixed a little bug in `clusterize_confiles.py`

Also: closes issue #280

### CODE REVIEW

**Comment by ttmc** [comment]
I changed things around a bit:
* `awsdeploy.sh` now has one optional argument: the name of the AWS deployment configuration file. That way, you can have a nice little set of AWS deployment config files and not be constantly editing `deploy_conf.py`.
* A default AWS deployment configuration file is provided, named `default_deploy_conf.py`. If you don't give `awsdeploy.sh` the name of an AWS deployment configuration file, then it will just use that file.
* I moved the validation code inside `launch_ec2_nodes.py`
* I added a check to see if the number of files in the `confiles` directory is equal to `NUM_NODES`. There's another check much later, but it's a problem that can be caught sooner, so now it is.
* I updated all the documentation accordingly.

**Comment by ttmc** [comment]
I changed my mind about having a default AWS deployment configuration file: I don't want people to accidentally deploy three m3.2xlarge instances on AWS and leave them running, because they might get a big unexpected bill. Therefore `awsdeploy.sh` now requires one command-line argument: the name of the AWS deployment configuration file. There is a helpful error message if they don't provide it.

I renamed `default_deploy_conf.py` to `example_deploy_conf.py`. That way users have a good starting point for their own AWS deployment configuration file.

I updated the documentation accordingly.

File: deploy-cluster-aws/launch_ec2_nodes.py [hunk 1]
```diff
@@ -36,14 +38,52 @@
...
+    WHAT_TO_DEPLOY = cf.WHAT_TO_DEPLOY
+    USE_KEYPAIRS_FILE = cf.USE_KEYPAIRS_FILE
+    IMAGE_ID = cf.IMAGE_ID
+    INSTANCE_TYPE = cf.INSTANCE_TYPE
+except AttributeError as e:
+    print('One of the AWS deployment configuration settings was '
+          'not set in the AWS deployment configuration file ' +
+          '{}'.format(deploy_conf_file))
+    print('Read this traceback to find out which one (in ALL_CAPS):')
+    raise
```
Comment (sbellem) [inline deploy-cluster-aws/launch_ec2_nodes.py]: perhaps you could add a custom exception such as:

```python
class ImproperlyConfigured(Exception):
    pass
```

and raise this exception with the relevant message

Comment (sbellem) [inline deploy-cluster-aws/launch_ec2_nodes.py]: what about using multiple `try/except` clauses, to pinpoint the exact problem, e.g.:

```python
err_msg = ('The AWS deployment configuration settings "{}" was not set in the '                                                                                                                                                                                                            
           'AWS ...')                                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                           
try:                                                                                                                                                                                                                                                                                       
    NUM_NODES = cf.NUM_NODES                                                                                                                                                                                                                                                               
except AttributeError:                                                                                                                                                                                                                                                                     
    raise ImproperlyConfigured(err_msg.format('NUM_NODES'))                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                           
try:                                                                                                                                                                                                                                                                                       
    BRANCH = cf.BRANCH                                                                                                                                                                                                                                                                     
except AttributeError:                                                                                                                                                                                                                                                                     
    raise ImproperlyConfigured(err_msg.format('BRANCH'))                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                           
try:                                                                                                                                                                                                                                                                                       
    WHAT_TO_DEPLOY = cf.WHAT_TO_DEPLOY                                                                                                                                                                                                                                                     
except AttributeError:                                                                                                                                                                                                                                                                     
    raise ImproperlyConfigured(err_msg.format('WHAT_TO_DEPLOY'))                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                           
try:                                                                                                                                                                                                                                                                                       
    USE_KEYPAIRS_FILE = cf.USE_KEYPAIRS_FILE                                                                                                                                                                                                                                               
except AttributeError:                                                                                                                                                                                                                                                                     
    raise ImproperlyConfigured(err_msg.format('USE_KEYPAIRS_FILE'))                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                           
try:                                                                                                                                                                                                                                                                                       
    IMAGE_ID = cf.IMAGE_ID                                                                                                                                                                                                                                                                 
except AttributeError:                                                                                                                                                                                                                                                                     
    raise ImproperlyConfigured(err_msg.format('IMAGE_ID'))                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                           
try:                                                                                                                                                                                                                                                                                       
    INSTANCE_TYPE = cf.INSTANCE_TYPE                                                                                                                                                                                                                                                       
except AttributeError:                                                                                                                                                                                                                                                                     
    raise ImproperlyConfigured(err_msg.format('INSTANCE_TYPE'))                                                                                         ```

File: deploy-cluster-aws/launch_ec2_nodes.py [hunk 2]
```diff
@@ -36,14 +38,52 @@
...
+    assert isinstance(NUM_NODES, int)
+    assert isinstance(BRANCH, str)
+    assert isinstance(WHAT_TO_DEPLOY, str)
+    assert isinstance(USE_KEYPAIRS_FILE, bool)
+    assert isinstance(IMAGE_ID, str)
+    assert isinstance(INSTANCE_TYPE, str)
+except AssertionError as e:
+    print('One of the AWS deployment settings has a value of the wrong type.')
+    print('Read this traceback to find out which one (in ALL_CAPS):')
+    raise
```
Comment (sbellem) [inline deploy-cluster-aws/launch_ec2_nodes.py]: similar remark as the above with respect to isolating the errors, and using a custom exception such as `ImproperlyConfigured`

**Comment by ttmc** [comment]
I made some changes to the validation code (and exception handling) as suggested by Sylvain.

**Comment by sbellem** [comment]
:+1:

**Comment by r-marques** [comment]
nice :+1:

**Pull request merged**